Add more admin endpoints and use grpc-gateway

cuongdo opened the pull request:
Resolves #4298

Apologies for the size of this change. It was not obvious that
grpc-gateway would work out until I implemented enough endpoints using
it.

These are the new endpoints (better docs are in server/admin.proto):
`/_admin/v1/users` - lists users
`/_admin/v1/databases/DATABASE/tables/TABLE` - grants, columns, and indexes for a table
`/_admin/v1/events` - returns most recent event log entries
`/_admin/v1/uidata` - sets and gets data for the admin UI

The shift to gRPC coincides with the overall shift of CockroachDB to
gRPC. To be clear, the admin UI will still access the UI through the
REST API.  This is done using grpc-gateway, which is a set of tools that
generates an HTTP reverse proxy that exposes REST-style HTTP endpoints
and proxies them to gRPC.

Though all endpoints are natively gRPC, we test them using their
REST-style JSON interface, because the admin UI will use the REST API
for at least the short-term. Because the admin UI doesn't have its own
tests (yet), we need to ensure the endpoints work in the way that the
admin UI uses them.

This change also replaces the specialized system.reporting table with
the more general-purpose system.ui table. We reuse system.reporting's
TableID, because that's reasonable pre-beta.

These changes necessitate some changes to 'make check'. To return
proper HTTP error codes (e.g. 404 Not Found), I need to use
grpc.Errorf, which has an error code as its first parameter. 'go vet'
doesn't like that the first parameter isn't a format string.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4844)

<!-- Reviewable:end -->

tamird commented:
This is nice, thanks for going down the grpc-gateway road. I'll finish this review tomorrow, but that this commit mixes refactoring of existing endpoints with the addition of new endpoints makes it quite hard to review, and that's a bummer.

---

Reviewed 6 of 14 files at r1.
Review status: 6 of 14 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[GLOCKFILE, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBydJKvmFe9eol5_KTC:-KBydJKwYYEA2sj7FmMw:252886851) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d30244d9ec994eab0856141fc1de2fadcb437e/GLOCKFILE#L56)):*
this went away because you ran `glock save` in OS X (rather than the build container); you'll need to bring it back.

---

*[Makefile, line 202 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByeT4SH9RIGvblwv0Z:-KByeT4SH9RIGvblwv0_:-173141357) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/Makefile#L202)):*
i think a comment here could be good (just what you have in the commit message)

---

*[build/protobuf.mk, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBye88mJAPvapljlipp:-KBye88mJAPvapljlipq:-857216503) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/build/protobuf.mk#L38)):*
nit: alignment

---

*[build/protobuf.mk, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByeAjGgTBfYDHSYod1:-KByeAjGgTBfYDHSYod2:-1961740709) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/build/protobuf.mk#L44)):*
can you put a comment on this? also looks like you could avoid some repetition here

---

*[build/protobuf.mk, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByeNVXOLZLj1bIhEeQ:-KByeNVXOLZLj1bIhEeR:678029372) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/build/protobuf.mk#L79)):*
sigh

---

*[server/admin.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByfPcQ-Ry52H0j0jQL:-KByfPcQ-Ry52H0j0jQM:-2115770686) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/server/admin.go#L111)):*
looks like this is now unused.

---

*[server/admin.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBygH_4B-S63a7-Ny89:-KBygH_4B-S63a7-Ny8A:-2030407097) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/server/admin.go#L114)):*
tiny nit: I'd remove the member names from these comments so that future renames don't accidentally rot the comment.

---

*[server/admin.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByggZyc8PynzACCILg:-KByggZyc8PynzACCILh:-105511338) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/server/admin.go#L144)):*
i think i'd prefer this initialization at construction time. also might be worth keeping a reference to the context on the adminServer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-:-KByh2EJOziYw857lvHZ:754402108)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 6 of 14 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1TkaJa8ZdyzrFj6sR:-KC1TkaJa8ZdyzrFj6sS:865379444) ([raw file](https://github.com/cockroachdb/cockroach/blob/981c0971e6fae09d70161aeacfbf66d07ec9843e/server/admin.go#L306)):*
This is so much easier if you use `gorp` or something similar which does all of the marshaling / unmarshaling for you if you supply a struct.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

Commit:
Add more admin endpoints and use grpc-gateway

Resolves #4298 and fixes #4859

Apologies for the size of this change. It was not obvious that
grpc-gateway would work out until I implemented enough endpoints using
it.

These are the new endpoints (better docs are in server/admin.proto):
/_admin/v1/users - lists users
/_admin/v1/databases/DATABASE/tables/TABLE - grants, columns, and
    indexes for a table
/_admin/v1/events - returns most recent event log entries
/_admin/v1/uidata - sets and gets data for the admin UI

The shift to gRPC coincides with the overall shift of CockroachDB to
gRPC. To be clear, the admin UI will still access the UI through the
REST API.  This is done using grpc-gateway, which is a set of tools that
generates an HTTP reverse proxy that exposes REST-style HTTP endpoints
and proxies them to gRPC.

Though all endpoints are natively gRPC, we test them using their
REST-style JSON interface, because the admin UI will use the REST API
for at least the short-term. Because the admin UI doesn't have its own
tests (yet), we need to ensure the endpoints work in the way that the
admin UI uses them.

These changes necessitate some changes to 'make check'. To return
proper HTTP error codes (e.g. 404 Not Found), I need to use
grpc.Errorf, which has an error code as its first parameter. 'go vet'
doesn't like that the first parameter isn't a format string.

cuongdo commented:
Addressed Tamir's initial feedback. `server/admin.pb.gw.go` has some changes, because I regenerated with a new version of `grpc-gateway` that fixes some bugs.

---

Review status: 6 of 14 files reviewed at latest revision, 9 unresolved discussions.

---

*[GLOCKFILE, line 56 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBydJKvmFe9eol5_KTC:-KC12QMiubq6ZtEmd6z3:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/78d30244d9ec994eab0856141fc1de2fadcb437e/GLOCKFILE#L56)):*
Done.

---

*[Makefile, line 202 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByeT4SH9RIGvblwv0Z:-KC13jvJnzG0SJjWrspw:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/Makefile#L202)):*
Done.

---

*[build/protobuf.mk, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBye88mJAPvapljlipp:-KC13rX_8nXPQVq7Su3Z:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/build/protobuf.mk#L38)):*
Done.

---

*[build/protobuf.mk, line 44 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByeAjGgTBfYDHSYod1:-KC14c83M9K1On0nHP3e:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/build/protobuf.mk#L44)):*
Done.

---

*[server/admin.go, line 111 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByfPcQ-Ry52H0j0jQL:-KC17r6J0kNbudVQLOZl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/server/admin.go#L111)):*
Done.

---

*[server/admin.go, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KBygH_4B-S63a7-Ny89:-KC18WKRA8Ex7FMTK51X:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/server/admin.go#L114)):*
Done.

---

*[server/admin.go, line 144 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KByggZyc8PynzACCILg:-KC1C6fOoEMh_EamZLAp:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/server/admin.go#L144)):*
Done.

---

*[server/admin.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1TkaJa8ZdyzrFj6sR:-KC1Uaf8BLtXGl0PwMIM:-838998073) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/server/admin.go#L306)):*
Is that possible here though? I'm using `sql.Executor` directly. I can't use the public `database/sql` interface, because the API will have its own authentication mechanism later.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-:-KC1WeLEySq8GZ6bDlRV:1330562967)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 14 files at r1, 13 of 13 files at r2.
Review status: 17 of 18 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[config/config_test.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1bDcKNJa-cTWSu_zV:-KC1bDcKNJa-cTWSu_zW:-699403279) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/config/config_test.go#L241)):*
why did this change? is it possible to make this code self documenting (i.e. devoid of magic numbers)?

---

*[server/admin.go, line 175 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZcfANbesi8fI5SPW:-KC1ZcfANbesi8fI5SPX:-1928538297) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L175)):*
how do you feel about embedding the `mux` and removing this function?

---

*[server/admin.go, line 204 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZmPvI_jKaZR6ep2N:-KC1ZmPvI_jKaZR6ep2O:449834503) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L204)):*
add the word TODO somewhere in here

---

*[server/admin.go, line 206 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZjNNtV6mkE2gINBC:-KC1ZjNNtV6mkE2gINBD:-1423368222) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L206)):*
s/req/_/

---

*[server/admin.go, line 213 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1Zu7YGRN3ueG-1S18:-KC1Zu7YGRN3ueG-1S19:1604118182) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L213)):*
possibly throughout: no need to call `Error()` here

---

*[server/admin.go, line 289 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_D51ZsAgQRFaDh6s:-KC1_D51ZsAgQRFaDh6t:-526855682) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L289)):*
does this not work with placeholders, or do we not have the necessary plumbing for passing the placeholders in?

---

*[server/admin.go, line 292 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_TSDNys0bb0PC5Qd:-KC1_TSEz2GC_guUB0nF:-1654312971) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L292)):*
possibly throughout: no need to call `String()` here

---

*[server/admin.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_bqHj5QcPUDdglQF:-KC1_bqHj5QcPUDdglQG:348690760) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L348)):*
ah, there's a TODO here. replicate it above, please, or move this code to a single location so that a single TODO can cover both.

---

*[server/admin.proto, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1a978-NZVVO0kNC5L:-KC1a978-NZVVO0kNC5M:-1110911912) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.proto#L93)):*
the duplication between this and the sql package's protos is unfortunate

---

*[server/admin_test.go, line 196 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1aIwGtZbtboaSrPoO:-KC1aIwGtZbtboaSrPoP:-1730081603) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin_test.go#L196)):*
possibly throughout: can you just use the proto?

---

*[server/admin_test.go, line 313 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1aUcropl8ogpDb-Pm:-KC1aUcropl8ogpDb-Pn:266186429) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin_test.go#L313)):*
all these can be the actual protos?

---

*[storage/client_split_test.go, line 644 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1amOsyOxqJLuwNSB7:-KC1amOsyOxqJLuwNSB8:-1493243438) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/storage/client_split_test.go#L644)):*
the change below seems to increase the size of `expKeys`, but this comment talks about time? what does one have to do with the other?

Also remove "two of these keys" - that's just asking for rot.

---

*[util/http.go, line 264 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1b4lL3kyu3B-3uYYp:-KC1b4lL3kyu3B-3uYYq:335191003) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/util/http.go#L264)):*
isn't there already a smarter function in this file that unmarshals json,protobuf, and yaml?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 10 of 18 files reviewed at latest revision, 14 unresolved discussions.

---

*[config/config_test.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1bDcKNJa-cTWSu_zV:-KC1fAbrMbfToWsxihos:-331914602) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/config/config_test.go#L241)):*
I added a reserved table. I've tried to make it clearer.

---

*[server/admin.go, line 175 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZcfANbesi8fI5SPW:-KC1fYMDf7--IMv44ezf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L175)):*
Done.

---

*[server/admin.go, line 204 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZmPvI_jKaZR6ep2N:-KC1fhWqQP8B3SBvScfX:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L204)):*
Done.

---

*[server/admin.go, line 206 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1ZjNNtV6mkE2gINBC:-KC1fkZDzQcjoZpW2wx_:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L206)):*
Done.

---

*[server/admin.go, line 213 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1Zu7YGRN3ueG-1S18:-KC1g5CZ1f4TOLeeZerk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L213)):*
Done.

---

*[server/admin.go, line 289 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_D51ZsAgQRFaDh6s:-KC1gG_bZwGL801la9b2:-1558843710) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L289)):*
`SHOW` statements don't support placeholders. I've made this clearer in the docs.

---

*[server/admin.go, line 292 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_TSDNys0bb0PC5Qd:-KC1huRFODc-ArWHYJzs:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L292)):*
Done.

---

*[server/admin.go, line 348 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1_bqHj5QcPUDdglQF:-KC1i0MCosmKKFfE_IVd:-203038815) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.go#L348)):*
I added a TODO above. I can't centralize this logic, because the previous instance deals only with a database name. This code deals with database & table.

---

*[server/admin.proto, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1a978-NZVVO0kNC5L:-KC1lFdjsiwO6UdKl7JW:-1405850437) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.proto#L93)):*
Yes. Is there any way around that though?

---

*[server/admin_test.go, line 196 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1aIwGtZbtboaSrPoO:-KC261bht6WCZEC9TBxn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin_test.go#L196)):*
Done.

---

*[server/admin_test.go, line 313 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1aUcropl8ogpDb-Pm:-KC25z3stMHnF1Mpiehe:-1044839718) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin_test.go#L313)):*
Yes! I just deleted a bunch of code.

---

*[storage/client_split_test.go, line 644 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1amOsyOxqJLuwNSB7:-KC1iZMd1_2oSyV3R-5X:704438934) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/storage/client_split_test.go#L644)):*
I've removed the comment. It made more sense to me when I was writing it than it does now.

---

*[util/http.go, line 264 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1b4lL3kyu3B-3uYYp:-KC1k4PJ9v0fqNRPy4KX:-526954350) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/util/http.go#L264)):*
I don't think so. `UnmarshalRequest` only unmarshals HTTP requests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 8 files at r3.
Review status: 17 of 18 files reviewed at latest revision, 6 unresolved discussions.

---

*[config/config_test.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1bDcKNJa-cTWSu_zV:-KC28AVELX9jFe5ZBqUJ:-854748837) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/config/config_test.go#L241)):*
Can you avoid the constant here? Seems like this is the value of `sql.MakeMetadataSchema.tables`. There is already the analogous `DescriptorCount` method on that structure that is used for exactly the same kind of test automation.

---

*[server/admin.go, line 306 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1TkaJa8ZdyzrFj6sR:-KC29LeDg2Apcx9z0uyM:-1261296378) ([raw file](https://github.com/cockroachdb/cockroach/blob/9224a8769c1070a0af7bb2dab7a01bda0ae38ced/server/admin.go#L306)):*
Spencer and I were talking about this at lunch - I think we should consider implementing a `database/sql/driver` that uses the Executor directly. I'll file an issue.

---

*[server/admin.proto, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1a978-NZVVO0kNC5L:-KC29YNPPr0KM0hZN3FO:-1352794961) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin.proto#L93)):*
I'm not sure, just calling it out. Perhaps other folks have ideas.

---

*[server/admin_test.go, line 313 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1aUcropl8ogpDb-Pm:-KC29mttpwpScl6XcKut:746934352) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/server/admin_test.go#L313)):*
Awesome.

---

*[storage/client_split_test.go, line 644 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC29vWprcVGZWQNMcNN:-KC29vWprcVGZWQNMcNO:175120472) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98c3809a12f9c25b38d6220fefedfb022c613b2/storage/client_split_test.go#L644)):*
see comment in `config_test.go` about this exact constant - I think we can automate it away.

---

*[util/http.go, line 264 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1b4lL3kyu3B-3uYYp:-KC2A4kueUrP6A3DC9Kj:157191470) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/util/http.go#L264)):*
Not sure what you mean - this function also seems to be http-specific. I'm suggesting that you use `MarshalResponse` and `UnmarshalRequest` to make {Get,Post}Json more general (they could deal with any of the content types supported by `MarshalResponse` and `UnmarshalRequest`).

Perhaps a TODO is sufficient, if you prefer.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 14 of 19 files reviewed at latest revision, 4 unresolved discussions.

---

*[config/config_test.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1bDcKNJa-cTWSu_zV:-KC2GZbgjGxi35Gq7vju:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/config/config_test.go#L241)):*
Done.

---

*[storage/client_split_test.go, line 644 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC29vWprcVGZWQNMcNN:-KC2GtNDFXovB1-kjXWN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/d98c3809a12f9c25b38d6220fefedfb022c613b2/storage/client_split_test.go#L644)):*
Done.

---

*[util/http.go, line 264 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1b4lL3kyu3B-3uYYp:-KC2HA1_84_wdQvdp8C2:-606513554) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/util/http.go#L264)):*
I'll add a TODO here. I'd rather not expand this PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM but at least @mrtracy should review this as well. Also I didn't review the reflection sorter.

---

Reviewed 4 of 4 files at r4.
Review status: 18 of 19 files reviewed at latest revision, 1 unresolved discussion.

---

*[config/config_test.go, line 241 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC1bDcKNJa-cTWSu_zV:-KC2UHOc5OcV888urZbQ:2057758214) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd8c59ee0d6c8b83d706656486d5bab5325c45a4/config/config_test.go#L241)):*
Sweet, thanks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-:-KC2UNxkidDnRBecEC-F:-1225075908)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 18 of 19 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 211 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC2aTc_YBg6FkJKhjVx:-KC2aTcaI2RFJ0TFWNUu:-225179564) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L211)):*
You could also do: `log.ErrorfDepth(1, "%s", err)`. Similar in `serverErrorf`.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KC3OADYEvYmln8Fx_Tt:-2007920501) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
Can you expand on this comment? Is this a bug we should fix?

---

*[server/admin.go, line 687 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OMh7QC8vyMsNWug4:-KC3OMh7QC8vyMsNWug5:-376662309) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L687)):*
Since you're using a different placeholder, why didn't you use a single character placeholder such as `$`? Seems like doing so would simplify `Append`.

---

*[server/admin.proto, line 116 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3PutPuf5NXyRlsJMm:-KC3PutPuf5NXyRlsJMn:994705194) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.proto#L116)):*
We should also return stats here such as number of ranges and number of bytes. Certainly not for this PR. 

---

*[server/admin.proto, line 224 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3PRggBE0-hN8d2VjJ:-KC3PRggBE0-hN8d2VjK:-1390020690) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.proto#L224)):*
`s/databases/users/g`

---

*[testutils/sort.go, line 46 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3QnxUC-UO7WRclO_G:-KC3QnxUC-UO7WRclO_H:-1741201459) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L46)):*
You could do this check in `SortStructs` so that you don't have to perform it on every call.

Then you could initialize `v1` and `v2` as:

```
v1 := reflect.Indirect(ss.v.Index(i))
v2 := reflect.Indirect(ss.v.Index(j))
```

See https://golang.org/src/reflect/value.go?s=61361:61389#L2065.

---

*[testutils/sort.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3ShnxZB21OgakfbGI:-KC3Shny29ROl6Suh9Dx:-327152618) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L60)):*
I think you can move this check into `SortStructs` as well. Something like:

```
v := reflect.ValueOf(s)
elemT := v.Type.Elem()
if elemT.Kind() == reflect.Ptr {
  elemT = elemT.Elem()
}
if elemT.Kind() != reflect.Struct {
  return err
}
f, ok := elemT.FieldByName(fieldName)
if !ok {
  return err
}
```

Note that you could then cache `f.Index` (i.e. `StructField.Index`) and use the more efficient `reflect.Value.FieldByIndex()` to find the field values here in `Less`. I don't think that is worth doing right now, but a TODO wouldn't hurt if anyone ever needs to optimize this code.

---

*[testutils/sort.go, line 68 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3Rsgu9EwYxG-J3C1b:-KC3Rsgu9EwYxG-J3C1c:2092015494) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L68)):*
Is there a reason you're casting the kinds to ints? You could use the `%s` specifier and output them as strings.

I'm not sure if this check is even necessary. I don't think it is possible for `k1` to not equal `k2`. Is there something I'm missing here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 18 of 19 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KC3exJIlvQDKLWjROHG:-284222685) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
I think it allows the below to begin with `AND` without special-casing the first `AND` to be a `WHERE`

---

*[server/admin.proto, line 224 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3PRggBE0-hN8d2VjJ:-KC3fD_nPpiGcfBi4CUF:-497043072) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.proto#L224)):*
I think these comments should just be removed in cases where they exactly duplicate the value of the option.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
haven't reviewed in detail, but :heart: the commit message.

petermattis commented:
Review status: 18 of 19 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KC7czTXS7OFPeoHOxbV:-655717796) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
Ah, that makes sense. The comment and the `1=1` confused me. Could this be `WHERE TRUE`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 11 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[server/admin.go, line 211 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC2aTc_YBg6FkJKhjVx:-KCGVumRC1UNM0QiB8I1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L211)):*
Done.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KCGi_j9Erg1WZ3I3Z1m:1019959526) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
Changed to `WHERE true` and changed the comment.

---

*[server/admin.go, line 687 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OMh7QC8vyMsNWug4:-KCGZ4wKeK0xj-v2cgop:1156547047) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L687)):*
I was trying for a combination of characters that was extra unlikely. However, `$` should work, so I've changed it.

---

*[server/admin.proto, line 116 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3PutPuf5NXyRlsJMm:-KCG_6gRZt67RhVHx4p-:387089995) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.proto#L116)):*
Filed an issue.

---

*[server/admin.proto, line 224 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3PRggBE0-hN8d2VjJ:-KCG_PXz3m4K-94S_7GY:515940378) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.proto#L224)):*
Corrected the comment.

---

*[testutils/sort.go, line 46 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3QnxUC-UO7WRclO_G:-KCGeB9ReyTYdltNl2hq:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L46)):*
Done.

---

*[testutils/sort.go, line 60 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3ShnxZB21OgakfbGI:-KCGfiWEqC7YluLRNVlG:-1352297949) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L60)):*
TODO added.

---

*[testutils/sort.go, line 68 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3Rsgu9EwYxG-J3C1b:-KCGgMs5HTdPzHuRwAtB:1268089880) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L68)):*
If one were to pass a `[]interface{}` into `SortStructs` with structs of different types but the same field name, this check could fail. However, I've improved and moved the check into `SortStructs`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@mrtracy, any feedback?

petermattis commented:
LGTM, though there is a tiny bit more simplification in `SortStructs`.

---

Review status: 11 of 19 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/admin.go, line 211 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC2aTc_YBg6FkJKhjVx:-KCGnIPuDd-1SgcE_opW:-527776075) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L211)):*
Hmm, I'm not seeing some of this change.

---

*[testutils/sort.go, line 68 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3Rsgu9EwYxG-J3C1b:-KCGmdImCBDMCsyb11UP:-700083688) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/testutils/sort.go#L68)):*
If you pass an `[]interface{}` to `SortStructs`, each element of the slice will have kind `reflect.Interface` which is neither `reflect.Ptr` or `reflect.Struct`.

---

*[testutils/sort.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KCGmwRzM0wAs7a9B3CD:-KCGmwS-_q21i8cJNF8Q:1195135114) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2070d6c09f884154e69d631eadaa94686c0acba/testutils/sort.go#L108)):*
Per my comment above, this is overkill. You only need to check that `structs.Elem()` is either a struct or a pointer to a struct. If an `[]interface{}` is passed to this method that check will fail. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-:-KCGnOglWIBeLsGXxYMx:-2035706483)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 8 files at r5.
Review status: 18 of 19 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/admin.go, line 211 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC2aTc_YBg6FkJKhjVx:-KCGnN1U91_IE3yeWrvl:1703351356) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L211)):*
Looks like publish-before-push.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KCGnREX4CQFtJCrQjLo:1703351356) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
Looks like publish-before-push.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 17 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[server/admin.go, line 211 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC2aTc_YBg6FkJKhjVx:-KCGpWR2N05mFXoLEpVY:-382839716) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L211)):*
Rebase/merge gone awry. Fixed.

---

*[server/admin.go, line 495 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KC3OADXQLBzO1ivH1Zk:-KCGp_BGEgpfE7FPQUap:-382839716) ([raw file](https://github.com/cockroachdb/cockroach/blob/33fbb06188293dc3e8f64898f6ab64d8b9ad08c6/server/admin.go#L495)):*
Rebase/merge gone awry. Fixed.

---

*[testutils/sort.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KCGmwRzM0wAs7a9B3CD:-KCGpTKj7efBoy1bnyGG:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2070d6c09f884154e69d631eadaa94686c0acba/testutils/sort.go#L108)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 17 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[testutils/sort.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KCGmwRzM0wAs7a9B3CD:-KCGqJmLdEZhMWSGvsf1:279737627) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2070d6c09f884154e69d631eadaa94686c0acba/testutils/sort.go#L108)):*
You're still checking every element of the struct. But in Go you know that all of the elements will be the same. So instead of looping you can get the `reflect.Type` using `structs.Elem()`:

```
  elemT := structs.Elem()
  if elemT.Kind() == reflect.Ptr {
    elemT = elemT.Elem()
  }
  if elemT.Kind() != reflect.Struct {
    panic(fmt.Sprintf("%s is not a struct or pointer to struct", structs.Elem()))
  }
```

You lose the check to see if pointers are nil, but I'd just let that panic during the actual sorting.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 17 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[testutils/sort.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KCGmwRzM0wAs7a9B3CD:-KCGuMoeldszKuGAEq-U:404886019) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2070d6c09f884154e69d631eadaa94686c0acba/testutils/sort.go#L108)):*
Ah, I didn't realize how much functionality was in the `reflect` package. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 17 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[testutils/sort.go, line 108 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844#-KCGmwRzM0wAs7a9B3CD:-KCGv1E-FgWHrHQC_SRQ:676834233) ([raw file](https://github.com/cockroachdb/cockroach/blob/e2070d6c09f884154e69d631eadaa94686c0acba/testutils/sort.go#L108)):*
LGTM

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r6.
Review status: 18 of 19 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4844)*
<!-- Sent from Reviewable.io -->
